pkg/agent/openflow/fields.go

```diff
@@ -102,20 +121,25 @@ var (
 	// reg4[0..18]: Field to store the union value of Endpoint port and Endpoint status. It is used as a single match
 	// when needed.
 	EpUnionField = binding.NewRegField(4, 0, 18, "EndpointUnion")
-	// reg4[19]: Field to mark that whether Service type is NodePort.
-	NodePortAddressField = binding.NewRegField(4, 19, 19, "NodePortAddress")
-	// ToNodePortAddressRegMark marks that the Service type as NodePort.
-	ToNodePortAddressRegMark = binding.NewRegMark(NodePortAddressField, 0b1)
-	// reg4[20]: Field to mark that whether the packet of Service NodePort/LoadBalancer from gateway requires SNAT.
-	ServiceSNATField = binding.NewRegField(4, 20, 20, "ServiceSNAT")
-	// ServiceNeedSNATRegMark marks that the packet of Service NodePort/LoadBalancer requires SNAT.
-	ServiceNeedSNATRegMark = binding.NewRegMark(ServiceSNATField, 0b1)
+	// reg4[19]: Mark to indicate the Service type is NodePort.
+	ToNodePortAddressRegMark = binding.NewOneBitRegMark(4, 19, "NodePortAddress")
 	// reg4[16..19]: Field to store the union value of Endpoint state and the mark of whether Service type is NodePort.
 	NodePortUnionField = binding.NewRegField(4, 16, 19, "NodePortUnion")
-	// reg4[21]: Mark to indicate the packet is from local AntreaFlexibleIPAM Pod.
-	// NotAntreaFlexibleIPAMRegMark will be used with RewriteMACRegMark, thus the reg id must not be same due to the limitation of ofnet library.
-	AntreaFlexibleIPAMRegMark    = binding.NewOneBitRegMark(4, 21, "AntreaFlexibleIPAM")
-	NotAntreaFlexibleIPAMRegMark = binding.NewOneBitZeroRegMark(4, 21, "AntreaFlexibleIPAM")
+	// reg4[20]: Field to indicate whether the packet is from local Antrea IPAM Pod. NotAntreaFlexibleIPAMRegMark will
+	// be used with RewriteMACRegMark, thus the reg id must not be same due to the limitation of ofnet library.
+	AntreaFlexibleIPAMRegMark    = binding.NewOneBitRegMark(4, 20, "AntreaFlexibleIPAM")
+	NotAntreaFlexibleIPAMRegMark = binding.NewOneBitZeroRegMark(4, 20, "NotAntreaFlexibleIPAM")
+	// reg4[21..22]: Field to store the state of the first packet of Service NodePort/LoadBalancer connection from Antrea
+	// gateway.
+	//  - 0b00: the packet doesn't require SNAT.
+	//  - 0b01: the packet requires SNAT and has not been marked with CT mark.
+	//  - 0b11: the packet requires SNAT and has been marked with CT mark.
+	ServiceSNATStateField = binding.NewRegField(4, 21, 22, "ServiceSNAT")
+	NotRequireSNATRegMark = binding.NewRegMark(ServiceSNATStateField, 0b00)
+	RequireSNATRegMark    = binding.NewRegMark(ServiceSNATStateField, 0b01)
+	CTMarkedSNATRegMark   = binding.NewRegMark(ServiceSNATStateField, 0b11)
+	// reg4[23]: Mark to indicate the packet is Egress packet.
+	EgressRegMark = binding.NewOneBitRegMark(4, 23, "Egress")
```

Comment:
Another concern from me is that the PR mixes framework change and flow changes. Take Egress as an example, I don't know we had the `EgressRegMark` mark before so I assume the way of identifying and matching egress traffic may have changed in this PR. However, even single match condition or action change could break several cases even when all tests pass. For example, in this PR https://github.com/antrea-io/antrea/pull/3371#discussion_r815856152, it seems adding a condition could fix one known bug but it also breaks two other cases. And I may have more questions like https://github.com/antrea-io/antrea/pull/3058#discussion_r814865225 when reviewing more flow changes.
The previous OpenFlow rules is definitely not perfect and must have unknown bugs, but it has been proved working for most cases after revised many times case by case. And each change went through a cautious review and discussion to ensure no regression, for example https://github.com/antrea-io/antrea/pull/2332/files#r661812942.
I'm not confident that a framework change along with many flow changes could fix all accumulated problems without introducing any regression. Maybe this PR could but hard for me to tell with these mixed changes. I would suggest to only make framework change in this PR and any flows that don't make sense or redundant should be updated via separated PRs for easy change identifying. But this is just my personal feeling and I'm fine with merging it in this way if other reviewers are confident about the flows changes.
@jianjuns @antoninbas @wenyingd @hongliangl

Reply:
@tnqn I agree with you. If there is a way to dissociate the framework changes from the flow changes, that would also be my preference. If this PR introduces any bug, it will be pretty hard to track down.
That being said, is it possible to change the framework in this PR without impacting the flow? I thought that some flow changes were inevitable due to the nature of the PR.

Reply:
@antoninbas thanks for your reply. Yes, I confirmed from @hongliangl that some flow changes have to be made due to the nature of the PR. He menioned only service and egress have such changes, and @wenyingd have reviewed the Service change, I will review the Egress changes. @hongliangl would try to revert other unnecessary flow changes.

Reply:
And flexible IPAM but that anyway is a new feature in alpha (but still @gran-vmv should review the changes carefully). I think we should consider what tests we can do for the pipeline changes to avoid regression, esp. for Service and Egress.